File: JanperCab/3-Application/Dtos/Enquiry/EnquiryListDto.cs
using _3_Application.Dtos.Customer;
using _3_Application.Dtos.Process;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace _3_Application.Dtos.Enquiry
{
    public class EnquiryListDto
    {
        public int Id { get; set; }

        public DateTime CreatedDate { get; set; }

        public int DaysInSystem
        {
            get
            {
                var dayDifference = (int)DateTime.Today.Subtract(CreatedDate).TotalDays;
                return Enumerable
                    .Range(1, dayDifference)
                    .Select(x => CreatedDate.AddDays(x))
                    .Count(x => x.DayOfWeek != DayOfWeek.Saturday && x.DayOfWeek != DayOfWeek.Sunday);
            }
        }

        public DateTime? LastEditted { get; set; }

        public DateTime? OrderedDate { get; set; }

        public DateTime? ApprovedDate { get; set; }

        public string CustomerReference { get; set; }

        public bool IsRoutingOnly { get; set; }

        public CustomerDto Customer { get; set; }

        public CustomerDto Manager { get; set; }

        public string DeliveryNote { get; set; }

        public bool HasBeenInvoiced { get; set; }



        public ICollection<ProcessDto> Processes { get; set; }



        public EnquiryListDto()
        {
            Processes = new Collection<ProcessDto>();
        }
    }
}

File: JanperCab/5-JanperCab/ClientApp/src/app/_models/enquiry/EnquiryListDto.ts
import { ManufacturerDto } from '../customer/ManufacturerDto';
import { CustomerDto } from 'src/app/_models/customer/CustomerDto';
import { DistributorDto } from '../customer/DistributorDto';
import { CabinetMakerDto } from 'src/app/_models/customer/CabinetMakerDto';
import { Type } from 'class-transformer';
import { ProcessCleaningDto } from '../process/ProcessCleaningDto';
import { ProcessDeliveringDto } from '../process/ProcessDeliveringDto';
import { ProcessPackingDto } from '../process/ProcessPackingDto';
import { ProcessPreRouteDto } from '../process/ProcessPreRouteDto';
import { ProcessPressingDto } from '../process/ProcessPressingDto';
import { ProcessRoutingDto } from '../process/ProcessRoutingDto';
import { ProcessDto } from '../process/ProcessDto';
import { ProcessTypeEnum } from 'src/app/_enums/ProcessTypeEnum';

export class EnquiryListDto {
  id: number;
  createdDate: Date;
  daysInSystem: number;
  lastEditted: Date;
  orderedDate: Date;
  approvedDate: Date;
  jobType: string;
  doorType: string;
  doorColour: string;
  customerReference: string;
  totalPrice: number;

  @Type(() => CustomerDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CabinetMakerDto,
          name: '_3_Application.Dtos.Customer.CabinetMakerDto, 3-Application',
        },
        {
          value: DistributorDto,
          name: '_3_Application.Dtos.Customer.DistributorDto, 3-Application',
        },
        {
          value: ManufacturerDto,
          name: '_3_Application.Dtos.Customer.ManufacturerDto, 3-Application',
        },
      ],
    },
  })
  customer: CustomerDto;

  @Type(() => CustomerDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CabinetMakerDto,
          name: '_3_Application.Dtos.Customer.CabinetMakerDto, 3-Application',
        },
        {
          value: DistributorDto,
          name: '_3_Application.Dtos.Customer.DistributorDto, 3-Application',
        },
        {
          value: ManufacturerDto,
          name: '_3_Application.Dtos.Customer.ManufacturerDto, 3-Application',
        },
      ],
    },
  })
  manager: CustomerDto;

  deliveryNote: string;
  hasBeenDelivered: boolean;
  hasBeenInvoiced: boolean;

  @Type(() => ProcessDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: ProcessPreRouteDto,
          name: '_3_Application.Dtos.Process.ProcessPreRouteDto, 3-Application',
        },
        {
          value: ProcessRoutingDto,
          name: '_3_Application.Dtos.Process.ProcessRoutingDto, 3-Application',
        },
        {
          value: ProcessPressingDto,
          name: '_3_Application.Dtos.Process.ProcessPressingDto, 3-Application',
        },
        {
          value: ProcessCleaningDto,
          name: '_3_Application.Dtos.Process.ProcessCleaningDto, 3-Application',
        },
        {
          value: ProcessPackingDto,
          name: '_3_Application.Dtos.Process.ProcessPackingDto, 3-Application',
        },
        {
          value: ProcessDeliveringDto,
          name: '_3_Application.Dtos.Process.ProcessDeliveringDto, 3-Application',
        },
      ],
    },
  })
  processes: ProcessDto[];

  get currentStatus(): ProcessDto {
    if (!this.processes) {
      return null;
    }

    const status = this.processes.find((x) => x.isCurrent);

    return status;
  }

  get statusDescription(): string {
    if (!this.currentStatus) {
      return 'Ordered';
    }

    return this.currentStatus.getStatus();
  }

  get isCompleted(): boolean {
    return this.hasBeenDelivered && this.hasBeenInvoiced;
  }
}
